src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -922,23 +1044,76 @@ - (void)_readThroughSkipped
                                      errorHandler:nil];
 }
 
-- (BOOL)_callDelegateWithBlock:(void (^)(id<MTRDeviceDelegate>))block
+- (BOOL)_delegateExists
 {
     os_unfair_lock_assert_owner(&self->_lock);
-    id<MTRDeviceDelegate> delegate = _weakDelegate.strongObject;
-    if (delegate) {
-        dispatch_async(_delegateQueue, ^{
-            block(delegate);
-        });
-        return YES;
+    for (MTRDeviceDelegateInfo * delegateInfo in _delegates) {
+        if (!delegateInfo.delegateIsNull) {
+            return YES;
+        }
     }
     return NO;
 }
 
+- (void)_iterateDelegatesWithBlock:(void (^)(MTRDeviceDelegateInfo * delegateInfo))block
+{
+    os_unfair_lock_assert_owner(&self->_lock);
+
+    if (!_delegates.count) {
+        MTR_LOG("%@ no delegates to iterate", self);
+    }
+
+    // Opportunistically remove defunct delegate references on every iteration
+    NSMutableSet * delegatesToRemove = [NSMutableSet set];
```

Comment:
Given this method is likely called quite often it would be good to avoid allocating this temporary container until we actually know we need to remove a delegate.